File: src/models/content.py
"""Unified Content model for all ingested content.

This module provides a unified Content model that replaces the separate
Newsletter and Document models. All content is stored with markdown as
the canonical format, aligning with parser output.

Content Sources:
- GMAIL: Email newsletters via Gmail API
- RSS: RSS/Atom feed articles
- FILE_UPLOAD: Uploaded documents (PDF, DOCX, etc.)
- YOUTUBE: YouTube video transcripts
- PODCAST: Podcast episode transcripts
- SUBSTACK: Substack newsletters via Substack API
- XSEARCH: X/Twitter posts via Grok API search
- BLOG: Blog posts scraped from index pages
- SCHOLAR: Academic papers via Semantic Scholar
- MANUAL: Manually created content via API
"""

from datetime import datetime
from enum import StrEnum
from typing import TYPE_CHECKING

from pydantic import BaseModel, ConfigDict, Field
from sqlalchemy import (
    JSON,
    Boolean,
    Column,
    DateTime,
    Enum as SQLEnum,
    ForeignKey,
    Index,
    Integer,
    String,
    Text,
)
from sqlalchemy.orm import Mapped, relationship

from src.models.base import Base

if TYPE_CHECKING:
    from src.models.image import Image
    from src.models.summary import Summary


class ContentSource(StrEnum):
    """Content source types.

    Unified enum replacing NewsletterSource with additional values
    for future content sources.
    """

    GMAIL = "gmail"
    RSS = "rss"
    FILE_UPLOAD = "file_upload"
    YOUTUBE = "youtube"
    PODCAST = "podcast"
    SUBSTACK = "substack"
    MANUAL = "manual"  # Manually created via API
    WEBPAGE = "webpage"  # Future: scraped web pages
    XSEARCH = "xsearch"  # X/Twitter posts via Grok search
    PERPLEXITY = "perplexity"  # Web articles via Perplexity search
    BLOG = "blog"  # Blog posts scraped from index pages
    SCHOLAR = "scholar"  # Academic papers via Semantic Scholar
    OTHER = "other"


class ContentStatus(StrEnum):
    """Processing status for content."""

    PENDING = "pending"  # Ingested, awaiting parsing
    PARSING = "parsing"  # Parser running
    PARSED = "parsed"  # Parsed, awaiting summarization
    PROCESSING = "processing"  # Summarization in progress
    COMPLETED = "completed"  # Fully processed
    FAILED = "failed"  # Processing failed


class Content(Base):  # type: ignore[valid-type, misc]
    """Unified content model for all ingested content.

    Stores content from all sources (Gmail, RSS, file uploads, YouTube)
    with markdown as the canonical format. This replaces the previous
    Newsletter + Document model combination.

    Key design decisions:
    - markdown_content is the primary content field (always populated)
    - raw_content preserves original format for re-parsing if needed
    - Structured data (tables, links) stored in JSON for querying
    - content_hash enables deduplication across all sources
    """

    __tablename__ = "contents"

    id = Column(Integer, primary_key=True, autoincrement=True)

    # Source identification
    # Note: values_callable ensures SQLAlchemy uses enum .value (lowercase)
    # instead of .name (uppercase) to match the database enum definition
    source_type = Column(
        SQLEnum(ContentSource, values_callable=lambda x: [e.value for e in x]),
        nullable=False,
    )
    source_id = Column(String(500), nullable=False)  # Unique ID from source
    source_url = Column(String(2000), nullable=True)  # Original URL if available

    # Identity / Metadata
    title = Column(String(1000), nullable=False)
    author = Column(String(500), nullable=True)  # Sender for email, channel for YouTube
    publication = Column(String(500), nullable=True)  # Newsletter/channel name
    published_date = Column(DateTime, nullable=True, index=True)

    # Canonical content - MARKDOWN FIRST
    markdown_content = Column(Text, nullable=False)  # Primary content, always populated

    # Structured extractions (parsed from markdown or source)
    tables_json = Column(JSON, nullable=True)  # List of TableData for complex tables
    links_json = Column(JSON, nullable=True)  # Extracted URLs for link analysis
    metadata_json = Column(
        JSON, nullable=True
    )  # Additional metadata (page_count, word_count, etc.)

    # Raw preservation (optional, for re-parsing)
    raw_content = Column(Text, nullable=True)  # Original HTML, transcript JSON, etc.
    raw_format = Column(String(50), nullable=True)  # "html", "text", "transcript_json"

    # Parsing metadata
    parser_used = Column(String(100), nullable=True)  # "DoclingParser", "YouTubeParser", etc.
    parser_version = Column(String(50), nullable=True)  # For tracking re-parsing needs

    # Deduplication
    content_hash = Column(String(64), nullable=False, index=True)  # SHA-256 of normalized markdown
    canonical_id = Column(
        Integer,
        ForeignKey("contents.id", ondelete="SET NULL"),
        nullable=True,
        index=True,
    )  # FK to canonical Content if duplicate

    # Processing status
    # Note: values_callable ensures SQLAlchemy uses enum .value (lowercase)
    # instead of .name (uppercase) to match the database enum definition
    status = Column(
        SQLEnum(ContentStatus, values_callable=lambda x: [e.value for e in x]),
        nullable=False,
        default=ContentStatus.PENDING,
        index=True,
    )
    error_message = Column(Text, nullable=True)

    # Sharing
    is_public = Column(Boolean, nullable=False, default=False, server_default="false")
    share_token = Column(String(36), nullable=True, unique=True, index=True)

    # Timestamps
    ingested_at = Column(DateTime, nullable=False, default=datetime.utcnow, index=True)
    parsed_at = Column(DateTime, nullable=True)
    processed_at = Column(DateTime, nullable=True)  # When summarization completed

    # Relationships
    canonical: Mapped["Content | None"] = relationship(
        "Content",
        remote_side="Content.id",
        foreign_keys=[canonical_id],
    )

    # Content → Summary relationship (one-to-many, typically one summary per content)
    summaries: Mapped[list["Summary"]] = relationship(
        "Summary",
        back_populates="content",
        foreign_keys="Summary.content_id",
    )

    # Content → Image relationship (one-to-many for extracted images)
    images: Mapped[list["Image"]] = relationship(
        "Image",
        back_populates="source_content",
        foreign_keys="Image.source_content_id",
    )

    # Content → DocumentChunk relationship (one-to-many for search chunks)
    chunks: Mapped[list["DocumentChunk"]] = relationship(  # type: ignore[name-defined]  # noqa: F821
        "DocumentChunk",
        back_populates="content",
        foreign_keys="DocumentChunk.content_id",
        cascade="all, delete-orphan",
        passive_deletes=True,
    )

    # Composite unique constraint on source_type + source_id
    __table_args__ = (
        Index("idx_contents_source", "source_type", "source_id", unique=True),
        Index("idx_contents_publication_date", "publication", "published_date"),
    )

    def __repr__(self) -> str:
        return f"<Content(id={self.id}, source={self.source_type.value}, title='{self.title[:50]}...')>"


# --- Pydantic Schemas ---


class ContentCreate(BaseModel):
    """Schema for creating new content."""

    source_type: ContentSource
    source_id: str
    source_url: str | None = None

    title: str
    author: str | None = None
    publication: str | None = None
    published_date: datetime | None = None

    markdown_content: str
    tables_json: list[dict] | None = None
    links_json: list[str] | None = None
    metadata_json: dict | None = None

    raw_content: str | None = None
    raw_format: str | None = None

    parser_used: str | None = None
    parser_version: str | None = None

    content_hash: str


class ContentUpdate(BaseModel):
    """Schema for updating content."""

    title: str | None = None
    author: str | None = None
    publication: str | None = None
    published_date: datetime | None = None

    markdown_content: str | None = None
    tables_json: list[dict] | None = None
    links_json: list[str] | None = None
    metadata_json: dict | None = None

    status: ContentStatus | None = None
    error_message: str | None = None


class ContentResponse(BaseModel):
    """Schema for content API responses."""

    id: int
    source_type: ContentSource
    source_id: str
    source_url: str | None = None

    title: str
    author: str | None = None
    publication: str | None = None
    published_date: datetime | None = None

    markdown_content: str
    tables_json: list[dict] | None = None
    links_json: list[str] | None = None
    metadata_json: dict | None = None

    parser_used: str | None = None
    content_hash: str
    canonical_id: int | None = None

    status: ContentStatus
    error_message: str | None = None

    is_public: bool = False
    share_token: str | None = None

    ingested_at: datetime
    parsed_at: datetime | None = None
    processed_at: datetime | None = None

    model_config = ConfigDict(from_attributes=True)


class ContentListItem(BaseModel):
    """Lightweight schema for listing content."""

    id: int
    source_type: ContentSource
    title: str
    publication: str | None = None
    published_date: datetime | None = None
    status: ContentStatus
    ingested_at: datetime

    model_config = ConfigDict(from_attributes=True)


class ContentListResponse(BaseModel):
    """Paginated content list response."""

    items: list[ContentListItem]
    total: int
    page: int
    page_size: int
    has_next: bool = Field(default=False)
    has_prev: bool = Field(default=False)


# --- Sharing Schemas ---


class ShareResponse(BaseModel):
    """Response for share enable/status endpoints."""

    is_public: bool
    share_token: str | None = None
    share_url: str | None = None

    model_config = ConfigDict(from_attributes=True)

File: web/src/types/content.ts
/**
 * Content Types
 *
 * TypeScript interfaces for the unified Content model.
 * These types mirror the backend Python models in src/models/content.py
 *
 * The Content model unifies newsletters, documents, and other content sources
 * into a single entity with markdown-first representation.
 *
 * @see Backend model: src/models/content.py
 */

/**
 * Source of the content
 * - gmail: Fetched from Gmail inbox
 * - rss: Fetched from RSS feed
 * - file_upload: Uploaded document (PDF, DOCX, etc.)
 * - youtube: YouTube video transcript
 * - manual: Manually created via API
 * - webpage: Scraped web page (future)
 * - other: Other sources
 */
export type ContentSource =
  | "gmail"
  | "rss"
  | "file_upload"
  | "youtube"
  | "manual"
  | "webpage"
  | "other"

/**
 * Processing status of content
 * - pending: Awaiting processing
 * - parsing: Being parsed to markdown
 * - parsed: Successfully parsed
 * - processing: Being summarized
 * - completed: Successfully processed
 * - failed: Processing encountered an error
 */
export type ContentStatus =
  | "pending"
  | "parsing"
  | "parsed"
  | "processing"
  | "completed"
  | "failed"

/**
 * Content entity
 *
 * Represents a single piece of content that has been ingested.
 * This is the unified model replacing Newsletter + Document.
 * Field names use snake_case to match the Python backend API responses.
 */
export interface Content {
  /** Unique identifier */
  id: number

  /** Source type of the content */
  source_type: ContentSource

  /** External ID from the source system */
  source_id: string

  /** Original URL if available */
  source_url: string | null

  /** Title of the content */
  title: string

  /** Author or sender */
  author: string | null

  /** Publication name if available */
  publication: string | null

  /** When the content was originally published */
  published_date: string | null // ISO 8601 date string

  /** Content in markdown format (primary content field) */
  markdown_content: string

  /** Structured table data extracted from content */
  tables_json: Record<string, unknown>[] | null

  /** URLs extracted from content */
  links_json: string[] | null

  /** Additional metadata */
  metadata_json: Record<string, unknown> | null

  /** Parser used to process the content */
  parser_used: string | null

  /** SHA-256 hash for deduplication */
  content_hash: string

  /** Canonical content ID if this is a duplicate */
  canonical_id: number | null

  /** Current processing status */
  status: ContentStatus

  /** Error message if status is 'failed' */
  error_message: string | null

  /** When the content was ingested into the system */
  ingested_at: string // ISO 8601 date string

  /** When parsing completed */
  parsed_at: string | null

  /** When processing completed */
  processed_at: string | null

  /** Legacy newsletter ID for navigating to summary review.
   * During migration, content is linked to newsletters via source_id.
   * This field provides direct access to the newsletter_id for summary lookups. */
  legacy_newsletter_id: number | null
}

/**
 * Content list item (summary view)
 *
 * Lightweight representation for list views.
 * Omits large fields like markdown_content.
 */
export interface ContentListItem {
  id: number
  source_type: ContentSource
  title: string
  publication: string | null
  published_date: string | null
  status: ContentStatus
  ingested_at: string
  /** Legacy newsletter ID for navigating to summary review */
  legacy_newsletter_id: number | null
}

/**
 * Paginated content list response
 */
export interface ContentListResponse {
  items: ContentListItem[]
  total: number
  page: number
  page_size: number
  has_next: boolean
  has_prev: boolean
}

/**
 * Sort order for table sorting
 */
export type SortOrder = "asc" | "desc"

/**
 * Filters for content list queries
 */
export interface ContentFilters {
  /** Filter by source type */
  source_type?: ContentSource
  /** Filter by processing status */
  status?: ContentStatus
  /** Filter by publication name */
  publication?: string
  /** Filter content after this date */
  start_date?: string
  /** Filter content before this date */
  end_date?: string
  /** Search in title */
  search?: string
  /** Page number (1-indexed) */
  page?: number
  /** Items per page */
  page_size?: number
  /** Field to sort by */
  sort_by?: string
  /** Sort direction */
  sort_order?: SortOrder
}

/**
 * Content statistics
 */
export interface ContentStats {
  total: number
  by_status: Record<ContentStatus, number>
  by_source: Record<ContentSource, number>
  pending_count: number
  completed_count: number
  failed_count: number
  /** Count of content items that don't have summaries yet */
  needs_summarization_count: number
}

/**
 * Request to create content via API
 */
export interface ContentCreateRequest {
  source_type?: ContentSource
  source_id?: string
  source_url?: string
  title: string
  author?: string
  publication?: string
  published_date?: string
  markdown_content: string
  tables_json?: Record<string, unknown>[]
  links_json?: string[]
  metadata_json?: Record<string, unknown>
}

/**
 * Duplicate content information
 */
export interface ContentDuplicateInfo {
  canonical_id: number
  canonical_title: string
  duplicate_count: number
}
